extensions/mongoProcessor/MongoQueueProcessor.js

```diff
@@ -392,43 +393,95 @@ class MongoQueueProcessor {
+     * @param {BucketInfo} bucketInfo - bucket info object
      * @param {function} done - callback(error)
      * @return {undefined}
      */
-    _processDeleteOpQueueEntry(log, sourceEntry, location, done) {
+    _processDeleteOpQueueEntry(log, sourceEntry, location, bucketInfo, done) {
         const bucket = sourceEntry.getBucket();
         const key = sourceEntry.getObjectKey();
-        const versionId = extractVersionId(sourceEntry.getObjectVersionedKey());
-
-        const options = versionId ? { versionId } : undefined;
-
-        // Calling deleteObject with undefined options to use deleteObjectNoVer which is used for
-        // deleting non versioned objects that only have master keys.
-        // When deleting a versioned object however we supply the version id in the options, which
-        // causes the function to call the deleteObjectVer function that is used to handle objects that
-        // have both a master and version keys. This handles the deletion of both the version and the master
-        // keys in the case where no other version is available, or deleting the version and updating the
-        // master key otherwise.
-        return this._mongoClient.deleteObject(bucket, key, options, log,
-            err => {
-                if (err) {
-                    this._normalizePendingMetric(location);
-                    log.end().error('error deleting object metadata ' +
-                    'from mongo', {
-                        bucket,
-                        key,
-                        error: err.message,
+        const entryVersionId = extractVersionId(sourceEntry.getObjectVersionedKey());
+
+        // Use x-amz-meta-scal-version-id if provided, instead of the actual versionId of the object.
+        // This should happen only for restored objects : in all other situations, both the source
+        // and ingested objects should have the same version id (and no x-amz-meta-scal-version-id
+        // metadata).
+        const scalVersionId = sourceEntry.getOverheadField('x-amz-meta-scal-version-id');
+        const versionId = scalVersionId ? VersionID.decode(scalVersionId) : entryVersionId;
+
+        this.logger.debug('processing object delete', { bucket, key, versionId });
+
+        async.waterfall([
+            cb => this._getZenkoObjectMetadata(log, sourceEntry, versionId, cb),
+            (zenkoObjMd, cb) => {
+                // Skip if the object is in a different location, i.e. when the delete was caused
+                // by restored-object expiration or transition. It works because the dataStoreName
+                // is updated before actually sending the object to GC to effectively delete the
+                // data.
+                const encode = versionId => (versionId ? VersionID.encode(versionId) : 'null');
+                if (zenkoObjMd.dataStoreName !== location ||
+                    zenkoObjMd.location?.length !== 1 ||
```

Comment:
> zenkoObjMd.location?.length !== 1
Why this check?
How about MPU with all its data parts stored at the same location?

Reply:
we want to skip the delete from OOB if the object is not in OOB : so this is mainly the `dataStoreName !== location` tests.

the extra conditions are sanity check, to ensure we "follow" OOB if the object is not in the expected state: an OOB object always has a single "part" (i.e. the s3c object!) which is also in the same location.

> How about MPU with all its data parts stored at the same location?

OOB locations are s3 locations, which is "mpu aware" : so the data is stored as a single multi-part object in the backend (i.e. in s3c), and the metadata only references a single object.